<!DOCTYPE html>
<!-- Generated by pkgdown: do not edit by hand --><html lang="en">
<head>
<meta http-equiv="Content-Type" content="text/html; charset=UTF-8">
<meta charset="utf-8">
<meta http-equiv="X-UA-Compatible" content="IE=edge">
<meta name="viewport" content="width=device-width, initial-scale=1, shrink-to-fit=no">
<title>Managing Mass Spectrometry Experiments • MsExperiment</title>
<script src="../deps/jquery-3.6.0/jquery-3.6.0.min.js"></script><meta name="viewport" content="width=device-width, initial-scale=1, shrink-to-fit=no">
<link href="../deps/bootstrap-5.3.1/bootstrap.min.css" rel="stylesheet">
<script src="../deps/bootstrap-5.3.1/bootstrap.bundle.min.js"></script><link href="../deps/font-awesome-6.5.2/css/all.min.css" rel="stylesheet">
<link href="../deps/font-awesome-6.5.2/css/v4-shims.min.css" rel="stylesheet">
<script src="../deps/headroom-0.11.0/headroom.min.js"></script><script src="../deps/headroom-0.11.0/jQuery.headroom.min.js"></script><script src="../deps/bootstrap-toc-1.0.1/bootstrap-toc.min.js"></script><script src="../deps/clipboard.js-2.0.11/clipboard.min.js"></script><script src="../deps/search-1.0.0/autocomplete.jquery.min.js"></script><script src="../deps/search-1.0.0/fuse.min.js"></script><script src="../deps/search-1.0.0/mark.min.js"></script><!-- pkgdown --><script src="../pkgdown.js"></script><meta property="og:title" content="Managing Mass Spectrometry Experiments">
</head>
<body>
    <a href="#main" class="visually-hidden-focusable">Skip to contents</a>


    <nav class="navbar navbar-expand-lg fixed-top bg-primary" data-bs-theme="dark" aria-label="Site navigation"><div class="container">

    <a class="navbar-brand me-2" href="../index.html">MsExperiment</a>

    <small class="nav-text text-muted me-auto" data-bs-toggle="tooltip" data-bs-placement="bottom" title="">1.11.1</small>


    <button class="navbar-toggler" type="button" data-bs-toggle="collapse" data-bs-target="#navbar" aria-controls="navbar" aria-expanded="false" aria-label="Toggle navigation">
      <span class="navbar-toggler-icon"></span>
    </button>

    <div id="navbar" class="collapse navbar-collapse ms-3">
      <ul class="navbar-nav me-auto">
<li class="active nav-item"><a class="nav-link" href="../articles/MsExperiment.html">Get started</a></li>
<li class="nav-item"><a class="nav-link" href="../reference/index.html">Reference</a></li>
<li class="nav-item"><a class="nav-link" href="../news/index.html">Changelog</a></li>
      </ul>
<ul class="navbar-nav">
<li class="nav-item"><form class="form-inline" role="search">
 <input class="form-control" type="search" name="search-input" id="search-input" autocomplete="off" aria-label="Search site" placeholder="Search for" data-search-index="../search.json">
</form></li>
<li class="nav-item"><a class="external-link nav-link" href="https://github.com/RforMassSpectrometry/MsExperiment/" aria-label="GitHub"><span class="fa fab fa-github fa-lg"></span></a></li>
      </ul>
</div>


  </div>
</nav><div class="container template-article">




<div class="row">
  <main id="main" class="col-md-9"><div class="page-header">
      <img src="../logo.png" class="logo" alt=""><h1>Managing Mass Spectrometry Experiments</h1>
            
      
      <small class="dont-index">Source: <a href="https://github.com/RforMassSpectrometry/MsExperiment/blob/main/vignettes/MsExperiment.Rmd" class="external-link"><code>vignettes/MsExperiment.Rmd</code></a></small>
      <div class="d-none name"><code>MsExperiment.Rmd</code></div>
    </div>

    
    
<script type="text/javascript">
document.addEventListener("DOMContentLoaded", function() {
  document.querySelector("h1").className = "title";
});
</script><script type="text/javascript">
document.addEventListener("DOMContentLoaded", function() {
  var links = document.links;  
  for (var i = 0, linksLength = links.length; i < linksLength; i++)
    if (links[i].hostname != window.location.hostname)
      links[i].target = '_blank';
});
</script><p><strong>Package</strong>: <em><a href="https://bioconductor.org/packages/3.22/MsExperiment" class="external-link">MsExperiment</a></em><br><strong>Authors</strong>: Laurent Gatto [aut, cre] (ORCID: <a href="https://orcid.org/0000-0002-1520-2268" class="external-link uri">https://orcid.org/0000-0002-1520-2268</a>), Johannes Rainer
[aut] (ORCID: <a href="https://orcid.org/0000-0002-6977-7147" class="external-link uri">https://orcid.org/0000-0002-6977-7147</a>), Sebastian Gibb
[aut] (ORCID: <a href="https://orcid.org/0000-0001-7406-4443" class="external-link uri">https://orcid.org/0000-0001-7406-4443</a>), Tuomas Borman
[ctb] (ORCID: <a href="https://orcid.org/0000-0002-8563-8884" class="external-link uri">https://orcid.org/0000-0002-8563-8884</a>)<br><strong>Last modified:</strong> 2025-06-04 06:30:40.349723<br><strong>Compiled</strong>: Wed Jun 4 06:39:37 2025</p>
<div class="section level2">
<h2 id="introduction">Introduction<a class="anchor" aria-label="anchor" href="#introduction"></a>
</h2>
<p>The goal of the <em><a href="https://bioconductor.org/packages/3.22/MsExperiment" class="external-link">MsExperiment</a></em>
package is to provide a container for all data related to a mass
spectrometry (MS) experiment. Also other Bioconductor packages allow to
represent MS experiment data (such as the <em><a href="https://bioconductor.org/packages/3.22/MSnbase" class="external-link">MSnbase</a></em>
package). The <code>MsExperiment</code> however aims at being very
light-weight and flexible to accommodate all possible types of MS
experiments (proteomics, metabolomics, …) and all types of MS data
representations (chromatographic and spectral data, quantified features
etc). In addition, it allows to bundle additional files and data, such
as annotations, within the object.</p>
<p>In this vignette, we will describe how to create a
<code>MsExperiment</code> object and populate it with various types of
data.</p>
<div class="sourceCode" id="cb1"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="st"><a href="https://github.com/RforMassSpectrometry/MsExperiment" class="external-link">"MsExperiment"</a></span><span class="op">)</span></span></code></pre></div>
<p>We will also use the <em><a href="https://bioconductor.org/packages/3.22/Spectra" class="external-link">Spectra</a></em>
package to import MS data and thus load it here too.</p>
<div class="sourceCode" id="cb2"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="st"><a href="https://github.com/RforMassSpectrometry/Spectra" class="external-link">"Spectra"</a></span><span class="op">)</span></span></code></pre></div>
</div>
<div class="section level2">
<h2 id="installation">Installation<a class="anchor" aria-label="anchor" href="#installation"></a>
</h2>
<p>The package can be installed with the <em><a href="https://bioconductor.org/packages/3.22/BiocManager" class="external-link">BiocManager</a></em>
package. To install <code>BiocManager</code> use
<code>install.packages("BiocManager")</code> and, after that,
<code>BiocManager::install("MsExperiment")</code> to install <em><a href="https://bioconductor.org/packages/3.22/MsExperiment" class="external-link">MsExperiment</a></em>
which will install the package including all required dependencies.</p>
</div>
<div class="section level2">
<h2 id="getting-data">Getting data<a class="anchor" aria-label="anchor" href="#getting-data"></a>
</h2>
<p>We will use a small subset of the <a href="https://www.ebi.ac.uk/pride/archive/projects/PXD022816" class="external-link">PXD022816</a>
project (<a href="https://doi.org/10.1021/acs.jproteome.0c00956" class="external-link">Morgenstern et
al. (2020)</a>). The acquisitions correspond to a Pierce Thermo HeLa
digestion standard, diluted to 50ng/uL with 97:3 + 0.1% formic acid, and
acquired on a QExactive instrument.</p>
<p>Below, we use the <em><a href="https://bioconductor.org/packages/3.22/rpx" class="external-link">rpx</a></em> package
to access the project from the PRIDE repository, and download files of
interest. Note that these will automatically be cached in the
<code>rpx</code> packages’ cache directory.</p>
<div class="sourceCode" id="cb3"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="st"><a href="https://github.com/lgatto/rpx" class="external-link">"rpx"</a></span><span class="op">)</span></span>
<span><span class="va">px</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/rpx/man/PXDataset2.html" class="external-link">PXDataset</a></span><span class="op">(</span><span class="st">"PXD022816"</span><span class="op">)</span></span></code></pre></div>
<pre><code><span><span class="co">## Loading PXD022816 from cache.</span></span></code></pre>
<div class="sourceCode" id="cb5"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">px</span></span></code></pre></div>
<pre><code><span><span class="co">## Project PXD022816 with 32 files</span></span>
<span><span class="co">## </span></span></code></pre>
<pre><code><span><span class="co">## Resource ID BFC1 in cache in /github/home/.cache/R/rpx.</span></span></code></pre>
<pre><code><span><span class="co">##  [1] 'QEP2LC6_HeLa_50ng_251120_01-calib.mzID.gz' ... [32] 'checksum.txt'</span></span>
<span><span class="co">##  Use 'pxfiles(.)' to see all files.</span></span></code></pre>
<div class="sourceCode" id="cb9"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/pkg/rpx/man/PXDataset2.html" class="external-link">pxfiles</a></span><span class="op">(</span><span class="va">px</span><span class="op">)</span></span></code></pre></div>
<pre><code><span><span class="co">## Project PXD022816 files (32):</span></span>
<span><span class="co">##  [remote] QEP2LC6_HeLa_50ng_251120_01-calib.mzID.gz</span></span>
<span><span class="co">##  [local]  QEP2LC6_HeLa_50ng_251120_01-calib.mzML</span></span>
<span><span class="co">##  [remote] QEP2LC6_HeLa_50ng_251120_01.raw</span></span>
<span><span class="co">##  [remote] QEP2LC6_HeLa_50ng_251120_02-calib.mzID.gz</span></span>
<span><span class="co">##  [local]  QEP2LC6_HeLa_50ng_251120_02-calib.mzML</span></span>
<span><span class="co">##  [remote] QEP2LC6_HeLa_50ng_251120_02.raw</span></span>
<span><span class="co">##  [remote] QEP2LC6_HeLa_50ng_251120_03-calib.mzID.gz</span></span>
<span><span class="co">##  [remote] QEP2LC6_HeLa_50ng_251120_03-calib.mzML</span></span>
<span><span class="co">##  [remote] QEP2LC6_HeLa_50ng_251120_03.raw</span></span>
<span><span class="co">##  [remote] QEP2LC6_HeLa_50ng_251120_04-calib.mzID.gz</span></span>
<span><span class="co">##  ...</span></span></code></pre>
<p>The project provides the vendor raw files, the converted mzML files
as well as the identification mzid files. Let’s download fractions 1 and
2 of the mzML files.</p>
<p>If you run these commands interactively and it’s the first time you
use <code><a href="https://rdrr.io/pkg/rpx/man/PXDataset2.html" class="external-link">pxget()</a></code>, you will be asked to create the
<code>rpx</code> cache directory - you can safelfy answer <em>yes</em>.
The files will then be downloaded. Next time you want to get the same
files, they will be loaded automatically from cache.</p>
<div class="sourceCode" id="cb11"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="op">(</span><span class="va">i</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/BiocGenerics/man/grep.html" class="external-link">grep</a></span><span class="op">(</span><span class="st">".+0[12].+mzML$"</span>, <span class="fu"><a href="https://rdrr.io/pkg/rpx/man/PXDataset2.html" class="external-link">pxfiles</a></span><span class="op">(</span><span class="va">px</span><span class="op">)</span>, value <span class="op">=</span> <span class="cn">TRUE</span><span class="op">)</span><span class="op">)</span></span></code></pre></div>
<pre><code><span><span class="co">## Project PXD022816 files (32):</span></span>
<span><span class="co">##  [remote] QEP2LC6_HeLa_50ng_251120_01-calib.mzID.gz</span></span>
<span><span class="co">##  [local]  QEP2LC6_HeLa_50ng_251120_01-calib.mzML</span></span>
<span><span class="co">##  [remote] QEP2LC6_HeLa_50ng_251120_01.raw</span></span>
<span><span class="co">##  [remote] QEP2LC6_HeLa_50ng_251120_02-calib.mzID.gz</span></span>
<span><span class="co">##  [local]  QEP2LC6_HeLa_50ng_251120_02-calib.mzML</span></span>
<span><span class="co">##  [remote] QEP2LC6_HeLa_50ng_251120_02.raw</span></span>
<span><span class="co">##  [remote] QEP2LC6_HeLa_50ng_251120_03-calib.mzID.gz</span></span>
<span><span class="co">##  [remote] QEP2LC6_HeLa_50ng_251120_03-calib.mzML</span></span>
<span><span class="co">##  [remote] QEP2LC6_HeLa_50ng_251120_03.raw</span></span>
<span><span class="co">##  [remote] QEP2LC6_HeLa_50ng_251120_04-calib.mzID.gz</span></span>
<span><span class="co">##  ...</span></span></code></pre>
<pre><code><span><span class="co">## [1] "QEP2LC6_HeLa_50ng_251120_01-calib.mzML"</span></span>
<span><span class="co">## [2] "QEP2LC6_HeLa_50ng_251120_02-calib.mzML"</span></span></code></pre>
<div class="sourceCode" id="cb14"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">fls</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/rpx/man/PXDataset2.html" class="external-link">pxget</a></span><span class="op">(</span><span class="va">px</span>, <span class="va">i</span><span class="op">)</span></span></code></pre></div>
<pre><code><span><span class="co">## Loading QEP2LC6_HeLa_50ng_251120_01-calib.mzML from cache.</span></span></code></pre>
<pre><code><span><span class="co">## Loading QEP2LC6_HeLa_50ng_251120_02-calib.mzML from cache.</span></span></code></pre>
<div class="sourceCode" id="cb17"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">fls</span></span></code></pre></div>
<pre><code><span><span class="co">## [1] "/github/home/.cache/R/rpx/d6c4c20495a_QEP2LC6_HeLa_50ng_251120_01-calib.mzML"</span></span>
<span><span class="co">## [2] "/github/home/.cache/R/rpx/d6c78e88909_QEP2LC6_HeLa_50ng_251120_02-calib.mzML"</span></span></code></pre>
</div>
<div class="section level2">
<h2 id="mass-spectrometry-experiment">Mass spectrometry experiment<a class="anchor" aria-label="anchor" href="#mass-spectrometry-experiment"></a>
</h2>
<p>Let’s start by creating an empty <code>MsExperiment</code> object
that we will populate with different pieces of data as we proceed with
the analysis of our data.</p>
<div class="sourceCode" id="cb19"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">msexp</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/MsExperiment.html">MsExperiment</a></span><span class="op">(</span><span class="op">)</span></span>
<span><span class="va">msexp</span></span></code></pre></div>
<pre><code><span><span class="co">## Empty object of class MsExperiment</span></span></code></pre>
<div class="section level3">
<h3 id="experiment-files">Experiment files<a class="anchor" aria-label="anchor" href="#experiment-files"></a>
</h3>
<p>Let’s now start with our MS experiment management by saving the
relevant files in a dedicated <code>MsExperimentFiles</code> object. In
addition to the mzML files, let’s also assume we have the human
proteomics fasta file ready. Later, when loading the raw data into R, we
will refer directly to the files in this <code>MsExperimentFiles</code>
object.</p>
<div class="sourceCode" id="cb21"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">msfls</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/MsExperimentFiles.html">MsExperimentFiles</a></span><span class="op">(</span>mzmls <span class="op">=</span> <span class="va">fls</span>,</span>
<span>                           fasta <span class="op">=</span> <span class="st">"homo_sapiens.fasta"</span><span class="op">)</span></span>
<span><span class="va">msfls</span></span></code></pre></div>
<pre><code><span><span class="co">## MsExperimentFiles of length  2 </span></span>
<span><span class="co">## [["mzmls"]] d6c4c20495a_QEP2LC6_HeLa_50ng_251120_01-calib.mzML ...</span></span>
<span><span class="co">## [["fasta"]] homo_sapiens.fasta</span></span></code></pre>
<p>Let’s add these files to the main experiment management object:</p>
<div class="sourceCode" id="cb23"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="../reference/MsExperiment.html">experimentFiles</a></span><span class="op">(</span><span class="va">msexp</span><span class="op">)</span> <span class="op">&lt;-</span> <span class="va">msfls</span></span>
<span><span class="va">msexp</span></span></code></pre></div>
<pre><code><span><span class="co">## Object of class MsExperiment </span></span>
<span><span class="co">##  Files: mzmls, fasta</span></span></code></pre>
</div>
<div class="section level3">
<h3 id="experimental-design">Experimental design<a class="anchor" aria-label="anchor" href="#experimental-design"></a>
</h3>
<p>The <code>sampleData</code> slot is used to describe the overall
experimental design of the experiment. It can be used to specify the
samples of the experiment and to relate them to the files that are part
of the experiment. There can be a one-to-one link between a sample and a
file, such as for example in label-free approaches, or one-to-many, in
labelled multiplexed approaches.</p>
<p>Here, we create a simple data frame with sample annotations that
include the original file names and the respective fractions.</p>
<div class="sourceCode" id="cb25"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="../reference/MsExperiment.html">sampleData</a></span><span class="op">(</span><span class="va">msexp</span><span class="op">)</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/S4Vectors/man/DataFrame-class.html" class="external-link">DataFrame</a></span><span class="op">(</span></span>
<span>    mzmls <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/pkg/BiocGenerics/man/path.html" class="external-link">basename</a></span><span class="op">(</span><span class="fu"><a href="../reference/MsExperiment.html">experimentFiles</a></span><span class="op">(</span><span class="va">msexp</span><span class="op">)</span><span class="op">[[</span><span class="st">"mzmls"</span><span class="op">]</span><span class="op">]</span><span class="op">)</span>,</span>
<span>    fractions <span class="op">=</span> <span class="fl">1</span><span class="op">:</span><span class="fl">2</span><span class="op">)</span></span>
<span><span class="fu"><a href="../reference/MsExperiment.html">sampleData</a></span><span class="op">(</span><span class="va">msexp</span><span class="op">)</span></span></code></pre></div>
<pre><code><span><span class="co">## DataFrame with 2 rows and 2 columns</span></span>
<span><span class="co">##           mzmls fractions</span></span>
<span><span class="co">##     &lt;character&gt; &lt;integer&gt;</span></span>
<span><span class="co">## 1 d6c4c20495...         1</span></span>
<span><span class="co">## 2 d6c78e8890...         2</span></span></code></pre>
</div>
<div class="section level3">
<h3 id="raw-data">Raw data<a class="anchor" aria-label="anchor" href="#raw-data"></a>
</h3>
<p>We can now create a <code>Spectra</code> object containing the raw
data stored in the mzML files. If you are not familiar with the
<code>Spectra</code> object, please refer to the <a href="https://rformassspectrometry.github.io/Spectra/articles/Spectra.html" class="external-link">package
vignettes</a>.</p>
<div class="sourceCode" id="cb27"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">sp</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/Spectra/man/Spectra.html" class="external-link">Spectra</a></span><span class="op">(</span><span class="fu"><a href="../reference/MsExperiment.html">experimentFiles</a></span><span class="op">(</span><span class="va">msexp</span><span class="op">)</span><span class="op">[[</span><span class="st">"mzmls"</span><span class="op">]</span><span class="op">]</span><span class="op">)</span></span>
<span><span class="va">sp</span></span></code></pre></div>
<pre><code><span><span class="co">## MSn data (Spectra) with 58907 spectra in a MsBackendMzR backend:</span></span>
<span><span class="co">##         msLevel     rtime scanIndex</span></span>
<span><span class="co">##       &lt;integer&gt; &lt;numeric&gt; &lt;integer&gt;</span></span>
<span><span class="co">## 1             1  0.177987         1</span></span>
<span><span class="co">## 2             1  0.599870         2</span></span>
<span><span class="co">## 3             1  0.978849         3</span></span>
<span><span class="co">## 4             1  1.363217         4</span></span>
<span><span class="co">## 5             1  1.742965         5</span></span>
<span><span class="co">## ...         ...       ...       ...</span></span>
<span><span class="co">## 58903         2   4198.59     29328</span></span>
<span><span class="co">## 58904         1   4198.74     29329</span></span>
<span><span class="co">## 58905         1   4199.11     29330</span></span>
<span><span class="co">## 58906         1   4199.49     29331</span></span>
<span><span class="co">## 58907         1   4199.87     29332</span></span>
<span><span class="co">##  ... 34 more variables/columns.</span></span>
<span><span class="co">## </span></span>
<span><span class="co">## file(s):</span></span>
<span><span class="co">## d6c4c20495a_QEP2LC6_HeLa_50ng_251120_01-calib.mzML</span></span>
<span><span class="co">## d6c78e88909_QEP2LC6_HeLa_50ng_251120_02-calib.mzML</span></span></code></pre>
<p>We can now add this object to the main experiment management
object:</p>
<div class="sourceCode" id="cb29"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/pkg/ProtGenerics/man/protgenerics.html" class="external-link">spectra</a></span><span class="op">(</span><span class="va">msexp</span><span class="op">)</span> <span class="op">&lt;-</span> <span class="va">sp</span></span>
<span><span class="va">msexp</span></span></code></pre></div>
<pre><code><span><span class="co">## Object of class MsExperiment </span></span>
<span><span class="co">##  Files: mzmls, fasta </span></span>
<span><span class="co">##  Spectra: MS1 (12983) MS2 (45924) </span></span>
<span><span class="co">##  Experiment data: 2 sample(s)</span></span></code></pre>
</div>
<div class="section level3">
<h3 id="third-party-applications">Third party applications<a class="anchor" aria-label="anchor" href="#third-party-applications"></a>
</h3>
<p>Let’s now assume we want to search the spectra in our mzML files
against the <code>homo_sapiens.fasta</code> file. To do so, we would
like to use a search engine such as MSGF+, that is run using the command
line and generates mzid files.</p>
<p>The command to run MSGF+ would look like this (see the <a href="https://msgfplus.github.io/msgfplus/MSGFPlus.html" class="external-link">manual page</a>
for details):</p>
<pre><code>java -jar /path/to/MSGFPlus.jar \
     -s input.mzML \
     -o output.mzid
     -d proteins.fasta \
     -t 20ppm \ ## precursor mass tolerance
     -tda 1 \   ## search decoy database
     -m 0 \     ## fragmentation method as written in the spectrum or CID if no info
     -int 1     ## Orbitrap/FTICR/Lumos</code></pre>
<p>We can easily build such a command for each of our input file:</p>
<div class="sourceCode" id="cb32"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">mzids</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/grep.html" class="external-link">sub</a></span><span class="op">(</span><span class="st">"mzML"</span>, <span class="st">"mzid"</span>, <span class="fu"><a href="https://rdrr.io/pkg/BiocGenerics/man/path.html" class="external-link">basename</a></span><span class="op">(</span><span class="fu"><a href="../reference/MsExperiment.html">experimentFiles</a></span><span class="op">(</span><span class="va">msexp</span><span class="op">)</span><span class="op">[[</span><span class="st">"mzmls"</span><span class="op">]</span><span class="op">]</span><span class="op">)</span><span class="op">)</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/base/paste.html" class="external-link">paste0</a></span><span class="op">(</span><span class="st">"java -jar /path/to/MSGFPlus.jar"</span>,</span>
<span>       <span class="st">" -s "</span>, <span class="fu"><a href="../reference/MsExperiment.html">experimentFiles</a></span><span class="op">(</span><span class="va">msexp</span><span class="op">)</span><span class="op">[[</span><span class="st">"mzmls"</span><span class="op">]</span><span class="op">]</span>,</span>
<span>       <span class="st">" -o "</span>, <span class="va">mzids</span>,</span>
<span>       <span class="st">" -d "</span>, <span class="fu"><a href="../reference/MsExperiment.html">experimentFiles</a></span><span class="op">(</span><span class="va">msexp</span><span class="op">)</span><span class="op">[[</span><span class="st">"fasta"</span><span class="op">]</span><span class="op">]</span>,</span>
<span>       <span class="st">" -t 20ppm"</span>,</span>
<span>       <span class="st">" -m 0"</span>,</span>
<span>       <span class="st">" int 1"</span><span class="op">)</span></span></code></pre></div>
<pre><code><span><span class="co">## [1] "java -jar /path/to/MSGFPlus.jar -s /github/home/.cache/R/rpx/d6c4c20495a_QEP2LC6_HeLa_50ng_251120_01-calib.mzML -o d6c4c20495a_QEP2LC6_HeLa_50ng_251120_01-calib.mzid -d homo_sapiens.fasta -t 20ppm -m 0 int 1"</span></span>
<span><span class="co">## [2] "java -jar /path/to/MSGFPlus.jar -s /github/home/.cache/R/rpx/d6c78e88909_QEP2LC6_HeLa_50ng_251120_02-calib.mzML -o d6c78e88909_QEP2LC6_HeLa_50ng_251120_02-calib.mzid -d homo_sapiens.fasta -t 20ppm -m 0 int 1"</span></span></code></pre>
<p>Here, for the sake of time and portability, we will not actually run
MSGF+, but a simple shell script that will generate mzid files in a
temporary R directory.</p>
<div class="sourceCode" id="cb34"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="op">(</span><span class="va">output</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/file.path.html" class="external-link">file.path</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/tempfile.html" class="external-link">tempdir</a></span><span class="op">(</span><span class="op">)</span>, <span class="va">mzids</span><span class="op">)</span><span class="op">)</span></span></code></pre></div>
<pre><code><span><span class="co">## [1] "/tmp/RtmpHXW8vs/d6c4c20495a_QEP2LC6_HeLa_50ng_251120_01-calib.mzid"</span></span>
<span><span class="co">## [2] "/tmp/RtmpHXW8vs/d6c78e88909_QEP2LC6_HeLa_50ng_251120_02-calib.mzid"</span></span></code></pre>
<div class="sourceCode" id="cb36"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">cmd</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/BiocGenerics/man/paste.html" class="external-link">paste</a></span><span class="op">(</span><span class="st">"touch"</span>, <span class="va">output</span><span class="op">)</span></span>
<span><span class="va">cmd</span></span></code></pre></div>
<pre><code><span><span class="co">## [1] "touch /tmp/RtmpHXW8vs/d6c4c20495a_QEP2LC6_HeLa_50ng_251120_01-calib.mzid"</span></span>
<span><span class="co">## [2] "touch /tmp/RtmpHXW8vs/d6c78e88909_QEP2LC6_HeLa_50ng_251120_02-calib.mzid"</span></span></code></pre>
<p>The <code>cmd</code> variable holds the two commands to be run on the
command line that will generate the new files. We can run each of these
commands with the <code><a href="https://rdrr.io/r/base/system.html" class="external-link">system()</a></code> function.</p>
<div class="sourceCode" id="cb38"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/pkg/BiocGenerics/man/lapply.html" class="external-link">sapply</a></span><span class="op">(</span><span class="va">cmd</span>, <span class="va">system</span><span class="op">)</span></span></code></pre></div>
<pre><code><span><span class="co">## touch /tmp/RtmpHXW8vs/d6c4c20495a_QEP2LC6_HeLa_50ng_251120_01-calib.mzid </span></span>
<span><span class="co">##                                                                        0 </span></span>
<span><span class="co">## touch /tmp/RtmpHXW8vs/d6c78e88909_QEP2LC6_HeLa_50ng_251120_02-calib.mzid </span></span>
<span><span class="co">##                                                                        0</span></span></code></pre>
<p>Below, we add the names of the newly created files to our
experiment:</p>
<div class="sourceCode" id="cb40"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="../reference/MsExperiment.html">experimentFiles</a></span><span class="op">(</span><span class="va">msexp</span><span class="op">)</span><span class="op">[[</span><span class="st">"mzids"</span><span class="op">]</span><span class="op">]</span> <span class="op">&lt;-</span> <span class="va">mzids</span></span>
<span><span class="fu"><a href="../reference/MsExperiment.html">experimentFiles</a></span><span class="op">(</span><span class="va">msexp</span><span class="op">)</span></span></code></pre></div>
<pre><code><span><span class="co">## MsExperimentFiles of length  3 </span></span>
<span><span class="co">## [["mzmls"]] d6c4c20495a_QEP2LC6_HeLa_50ng_251120_01-calib.mzML ...</span></span>
<span><span class="co">## [["fasta"]] homo_sapiens.fasta</span></span>
<span><span class="co">## [["mzids"]] d6c4c20495a_QEP2LC6_HeLa_50ng_251120_01-calib.mzid ...</span></span></code></pre>
<div class="sourceCode" id="cb42"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">msexp</span></span></code></pre></div>
<pre><code><span><span class="co">## Object of class MsExperiment </span></span>
<span><span class="co">##  Files: mzmls, fasta, mzids </span></span>
<span><span class="co">##  Spectra: MS1 (12983) MS2 (45924) </span></span>
<span><span class="co">##  Experiment data: 2 sample(s)</span></span></code></pre>
<p>We can also decide to store the commands that were used to generate
the mzid files in the experiment’s metadata slot. Here, we use the
convention to name that metadata item <code>"mzmls_to_mzids"</code> to
document to input and output of these commands.</p>
<div class="sourceCode" id="cb44"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/pkg/S4Vectors/man/Annotated-class.html" class="external-link">metadata</a></span><span class="op">(</span><span class="va">msexp</span><span class="op">)</span><span class="op">[[</span><span class="st">"mzmls_to_mzids"</span><span class="op">]</span><span class="op">]</span> <span class="op">&lt;-</span> <span class="va">cmd</span></span>
<span><span class="fu"><a href="https://rdrr.io/pkg/S4Vectors/man/Annotated-class.html" class="external-link">metadata</a></span><span class="op">(</span><span class="va">msexp</span><span class="op">)</span></span></code></pre></div>
<pre><code><span><span class="co">## $mzmls_to_mzids</span></span>
<span><span class="co">## [1] "touch /tmp/RtmpHXW8vs/d6c4c20495a_QEP2LC6_HeLa_50ng_251120_01-calib.mzid"</span></span>
<span><span class="co">## [2] "touch /tmp/RtmpHXW8vs/d6c78e88909_QEP2LC6_HeLa_50ng_251120_02-calib.mzid"</span></span></code></pre>
<p>Finally, the <code><a href="../reference/MsExperimentFiles.html">existMsExperimentFiles()</a></code> can be used at any
time to check which of files that are associated with an experiment
actually exist:</p>
<div class="sourceCode" id="cb46"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="../reference/MsExperimentFiles.html">existMsExperimentFiles</a></span><span class="op">(</span><span class="va">msexp</span><span class="op">)</span></span></code></pre></div>
<pre><code><span><span class="co">## mzmls: 2 out of 2 exist(s)</span></span>
<span><span class="co">## fasta: 0 out of 1 exist(s)</span></span>
<span><span class="co">## mzids: 0 out of 2 exist(s)</span></span></code></pre>
</div>
</div>
<div class="section level2">
<h2 id="saving-and-reusing-experiments">Saving and reusing experiments<a class="anchor" aria-label="anchor" href="#saving-and-reusing-experiments"></a>
</h2>
<p>The <code>MsExperiment</code> object has been used to store files and
data pertaining to a mass spectrometry experiment. It is now possible to
save that object and reload it later to recover all data and metadata.
See also section <em>Using <code>MsExperiment</code> with
<code>MsBackendSql</code></em> below for an alternative to load/restore
defined MS experiments from an SQL database.</p>
<div class="sourceCode" id="cb48"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/pkg/BiocGenerics/man/saveRDS.html" class="external-link">saveRDS</a></span><span class="op">(</span><span class="va">msexp</span>, <span class="st">"msexp.rds"</span><span class="op">)</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/base/rm.html" class="external-link">rm</a></span><span class="op">(</span>list <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/ls.html" class="external-link">ls</a></span><span class="op">(</span><span class="op">)</span><span class="op">)</span></span></code></pre></div>
<div class="sourceCode" id="cb49"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">msexp</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/readRDS.html" class="external-link">readRDS</a></span><span class="op">(</span><span class="st">"msexp.rds"</span><span class="op">)</span></span>
<span><span class="va">msexp</span></span></code></pre></div>
<pre><code><span><span class="co">## Object of class MsExperiment </span></span>
<span><span class="co">##  Files: mzmls, fasta, mzids </span></span>
<span><span class="co">##  Spectra: MS1 (12983) MS2 (45924) </span></span>
<span><span class="co">##  Experiment data: 2 sample(s)</span></span></code></pre>
<div class="sourceCode" id="cb51"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="../reference/MsExperiment.html">experimentFiles</a></span><span class="op">(</span><span class="va">msexp</span><span class="op">)</span></span></code></pre></div>
<pre><code><span><span class="co">## MsExperimentFiles of length  3 </span></span>
<span><span class="co">## [["mzmls"]] d6c4c20495a_QEP2LC6_HeLa_50ng_251120_01-calib.mzML ...</span></span>
<span><span class="co">## [["fasta"]] homo_sapiens.fasta</span></span>
<span><span class="co">## [["mzids"]] d6c4c20495a_QEP2LC6_HeLa_50ng_251120_01-calib.mzid ...</span></span></code></pre>
<p>We can access the raw data as long as the mzML files that were used
to generate it still exist in their original location, which is the case
here as they were saved in the <code>rpx</code> cache directory.</p>
<div class="sourceCode" id="cb53"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">sp</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/ProtGenerics/man/protgenerics.html" class="external-link">spectra</a></span><span class="op">(</span><span class="va">msexp</span><span class="op">)</span></span>
<span><span class="va">sp</span></span></code></pre></div>
<pre><code><span><span class="co">## MSn data (Spectra) with 58907 spectra in a MsBackendMzR backend:</span></span>
<span><span class="co">##         msLevel     rtime scanIndex</span></span>
<span><span class="co">##       &lt;integer&gt; &lt;numeric&gt; &lt;integer&gt;</span></span>
<span><span class="co">## 1             1  0.177987         1</span></span>
<span><span class="co">## 2             1  0.599870         2</span></span>
<span><span class="co">## 3             1  0.978849         3</span></span>
<span><span class="co">## 4             1  1.363217         4</span></span>
<span><span class="co">## 5             1  1.742965         5</span></span>
<span><span class="co">## ...         ...       ...       ...</span></span>
<span><span class="co">## 58903         2   4198.59     29328</span></span>
<span><span class="co">## 58904         1   4198.74     29329</span></span>
<span><span class="co">## 58905         1   4199.11     29330</span></span>
<span><span class="co">## 58906         1   4199.49     29331</span></span>
<span><span class="co">## 58907         1   4199.87     29332</span></span>
<span><span class="co">##  ... 34 more variables/columns.</span></span>
<span><span class="co">## </span></span>
<span><span class="co">## file(s):</span></span>
<span><span class="co">## d6c4c20495a_QEP2LC6_HeLa_50ng_251120_01-calib.mzML</span></span>
<span><span class="co">## d6c78e88909_QEP2LC6_HeLa_50ng_251120_02-calib.mzML</span></span></code></pre>
<div class="sourceCode" id="cb55"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/pkg/Spectra/man/spectra-plotting.html" class="external-link">plotSpectra</a></span><span class="op">(</span><span class="va">sp</span><span class="op">[</span><span class="fl">1000</span><span class="op">]</span><span class="op">)</span></span></code></pre></div>
<p><img src="MsExperiment_files/figure-html/plot_sp-1.png" width="700"></p>
</div>
<div class="section level2">
<h2 id="linking-experimental-data-to-samples">Linking experimental data to samples<a class="anchor" aria-label="anchor" href="#linking-experimental-data-to-samples"></a>
</h2>
<p>For some experiments and data analyses an explicit link between data,
data files and respective samples is required. Such links enable an easy
(and error-free) subset or re-ordering of a whole experiment by sample
and would also simplify coloring and labeling of the data depending on
the sample or of its variables or conditions.</p>
<p>Below we generate an <code>MsExperiment</code> object for a simple
experiment consisting of a single sample measured in two different
injections to the same LC-MS setup.</p>
<div class="sourceCode" id="cb56"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">lmse</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/MsExperiment.html">MsExperiment</a></span><span class="op">(</span><span class="op">)</span></span>
<span><span class="va">sd</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/S4Vectors/man/DataFrame-class.html" class="external-link">DataFrame</a></span><span class="op">(</span>sample_id <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"QC1"</span>, <span class="st">"QC2"</span><span class="op">)</span>,</span>
<span>                sample_name <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"QC Pool"</span>, <span class="st">"QC Pool"</span><span class="op">)</span>,</span>
<span>                injection_idx <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="fl">1</span>, <span class="fl">3</span><span class="op">)</span><span class="op">)</span></span>
<span><span class="fu"><a href="../reference/MsExperiment.html">sampleData</a></span><span class="op">(</span><span class="va">lmse</span><span class="op">)</span> <span class="op">&lt;-</span> <span class="va">sd</span></span></code></pre></div>
<p>We next add mzML files to the experiment for the sample that was
measured. These are available within the <code>msdata</code> R package.
We add also an additional <em>annotation</em> file
<code>"internal_standards.txt"</code> to the experiment, which could be
e.g. a file with m/z and retention times of internal standards added to
the sample (note that such files don’t necessarily have to exist).</p>
<div class="sourceCode" id="cb57"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">fls</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/list.files.html" class="external-link">dir</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/system.file.html" class="external-link">system.file</a></span><span class="op">(</span><span class="st">"sciex"</span>, package <span class="op">=</span> <span class="st">"msdata"</span><span class="op">)</span>, full.names <span class="op">=</span> <span class="cn">TRUE</span><span class="op">)</span></span>
<span><span class="fu"><a href="https://rdrr.io/pkg/BiocGenerics/man/path.html" class="external-link">basename</a></span><span class="op">(</span><span class="va">fls</span><span class="op">)</span></span></code></pre></div>
<pre><code><span><span class="co">## [1] "20171016_POOL_POS_1_105-134.mzML" "20171016_POOL_POS_3_105-134.mzML"</span></span></code></pre>
<div class="sourceCode" id="cb59"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="../reference/MsExperiment.html">experimentFiles</a></span><span class="op">(</span><span class="va">lmse</span><span class="op">)</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/MsExperimentFiles.html">MsExperimentFiles</a></span><span class="op">(</span></span>
<span>    mzML_files <span class="op">=</span> <span class="va">fls</span>,</span>
<span>    annotations <span class="op">=</span> <span class="st">"internal_standards.txt"</span><span class="op">)</span></span></code></pre></div>
<p>Next we load the MS data from the mzML files as a
<code>Spectra</code> object and add them to the experiment (see the
vignette of the <em><a href="https://bioconductor.org/packages/3.22/Spectra" class="external-link">Spectra</a></em>
for details on import and representation of MS data).</p>
<div class="sourceCode" id="cb60"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">sps</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/Spectra/man/Spectra.html" class="external-link">Spectra</a></span><span class="op">(</span><span class="va">fls</span>, backend <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/pkg/Spectra/man/MsBackend.html" class="external-link">MsBackendMzR</a></span><span class="op">(</span><span class="op">)</span><span class="op">)</span></span>
<span><span class="fu"><a href="https://rdrr.io/pkg/ProtGenerics/man/protgenerics.html" class="external-link">spectra</a></span><span class="op">(</span><span class="va">lmse</span><span class="op">)</span> <span class="op">&lt;-</span> <span class="va">sps</span></span>
<span><span class="va">lmse</span></span></code></pre></div>
<pre><code><span><span class="co">## Object of class MsExperiment </span></span>
<span><span class="co">##  Files: mzML_files, annotations </span></span>
<span><span class="co">##  Spectra: MS1 (1862) </span></span>
<span><span class="co">##  Experiment data: 2 sample(s)</span></span></code></pre>
<p>At this stage we have thus sample annotations and MS data in our
object, but no explicit relationships between them. Without such linking
between files and samples any subsetting would only subset the
<code>sampleData</code> but not any of the potentially associated files.
We next use the <code>linkSpectraData</code> function to establish and
define such relationships. First we link the experimental files to the
samples: we want to link the first mzML file in the element called
<code>"mzML_file"</code> in the object’s <code>experimentFiles</code> to
the first row in <code>sampleData</code> and the second file to the
second row.</p>
<div class="sourceCode" id="cb62"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">lmse</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/MsExperiment.html">linkSampleData</a></span><span class="op">(</span><span class="va">lmse</span>, with <span class="op">=</span> <span class="st">"experimentFiles.mzML_file"</span>,</span>
<span>                        sampleIndex <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="fl">1</span>, <span class="fl">2</span><span class="op">)</span>, withIndex <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="fl">1</span>, <span class="fl">2</span><span class="op">)</span><span class="op">)</span></span></code></pre></div>
<p>To define the link we have thus to specify with which
<em>element</em> within our <code>MsExperiment</code> we want to link
samples. This can be done with the parameter <code>with</code> that
takes a single <code>character</code> representing the name
(<em>address</em>) of the data element. The name is a combination of the
name of the slot within the <code>MsExperiment</code> and the name of
the element (or column) within that slot separated by a
<code>"."</code>. Using <code>with = "experimentFiles.mzML_file"</code>
means we want to link samples to values within the
<code>"mzML_file"</code> element of the object’s
<code>experimentFiles</code> slot - in other words, we want to link
samples to values in <code>experimentFiles(lmse)$mzML_file</code>. The
indices of the rows (samples) in <code>sampleData</code> and the indices
of the values in <code>with</code> to which we want to link the samples
can be defined with <code>sampleIndex</code> and <code>withIndex</code>.
In the example above we used <code>sampleIndex = c(1, 2)</code> and
<code>withIndex = c(1, 2)</code>, thus, we want to link the first row in
<code>sampleData</code> to the first value in <code>with</code> and the
second row to the second value. See also the section <em>Linking sample
data to other experimental data</em> in the documentation of
<code>MsExperiment</code> for more information and details.</p>
<p>What happened internally by the call above is illustrated in the
figure below. The link is represented as a two-column integer
<code>matrix</code> with the indices of the linked sample in the first
and the indices of the associated elements in the second columns (this
matrix is essentially a <code>cbind(sampleIndex, withIndex)</code>).</p>
<p><img src="imgs/Links_01.png"><!-- --></p>
<p>We next establish a second link between each sample and the
<em>annotation</em> file <code>"internal_standards.txt"</code> in
<code>experimentFiles(lmse)$standards</code>:</p>
<div class="sourceCode" id="cb63"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">lmse</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/MsExperiment.html">linkSampleData</a></span><span class="op">(</span><span class="va">lmse</span>, with <span class="op">=</span> <span class="st">"experimentFiles.annotations"</span>,</span>
<span>                        sampleIndex <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="fl">1</span>, <span class="fl">2</span><span class="op">)</span>, withIndex <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="fl">1</span>, <span class="fl">1</span><span class="op">)</span><span class="op">)</span></span></code></pre></div>
<p>The figure below illustrates again what happened internally by this
call: a new <em>link matrix</em> was added establishing the relationship
between the two samples and the one value in
<code>experimentFiles(lmse)$annotations</code>.</p>
<p><img src="imgs/Links_02.png"><!-- --></p>
<p>It is thus also possible to link different samples to the same
element. We next link the spectra in the object to the individual
samples. We use for that an alternative way to specify the link without
the need to provide <code>sampleIndex</code> and <code>withIndex</code>.
Sample-to-data links can also be specified using a syntax similar to an
SQL join:
<code>"sampleData.&lt;column in sampleData&gt; = &lt;slot&gt;.&lt;element in slot&gt;"</code>.
Links will be thus established between elements with matching values in
the specified data fields (i.e. between rows in <code>sampleData</code>
for which values in the specified column matches values in
<code>&lt;slot&gt;.&lt;element&gt;</code>). In order to use this
alternative approach to link spectra to the respective samples we have
to first add the (full) raw file name as an additional column to the
object’s <code>sampleData</code>. We can now add links between spectra
and samples by matching this raw file name to the original file name
from which the spectra were imported (which is available in the
<code>"dataOrigin"</code> spectra variable).</p>
<div class="sourceCode" id="cb64"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="../reference/MsExperiment.html">sampleData</a></span><span class="op">(</span><span class="va">lmse</span><span class="op">)</span><span class="op">$</span><span class="va">raw_file</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/normalizePath.html" class="external-link">normalizePath</a></span><span class="op">(</span><span class="va">fls</span><span class="op">)</span></span>
<span><span class="va">lmse</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/MsExperiment.html">linkSampleData</a></span><span class="op">(</span></span>
<span>    <span class="va">lmse</span>, with <span class="op">=</span> <span class="st">"sampleData.raw_file = spectra.dataOrigin"</span><span class="op">)</span></span></code></pre></div>
<p>The link was thus established between matching values in
<code>sampleData(lmse)$raw_file</code> and
<code>spectra(lmse)$dataOrigin</code>.</p>
<div class="sourceCode" id="cb65"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="../reference/MsExperiment.html">sampleData</a></span><span class="op">(</span><span class="va">lmse</span><span class="op">)</span><span class="op">$</span><span class="va">raw_file</span></span></code></pre></div>
<pre><code><span><span class="co">## [1] "/__w/_temp/Library/msdata/sciex/20171016_POOL_POS_1_105-134.mzML"</span></span>
<span><span class="co">## [2] "/__w/_temp/Library/msdata/sciex/20171016_POOL_POS_3_105-134.mzML"</span></span></code></pre>
<div class="sourceCode" id="cb67"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/pkg/ProtGenerics/man/protgenerics.html" class="external-link">spectra</a></span><span class="op">(</span><span class="va">lmse</span><span class="op">)</span><span class="op">$</span><span class="va">dataOrigin</span> <span class="op">|&gt;</span></span>
<span>                <span class="fu"><a href="https://rdrr.io/r/utils/head.html" class="external-link">head</a></span><span class="op">(</span><span class="op">)</span></span></code></pre></div>
<pre><code><span><span class="co">## [1] "/__w/_temp/Library/msdata/sciex/20171016_POOL_POS_1_105-134.mzML"</span></span>
<span><span class="co">## [2] "/__w/_temp/Library/msdata/sciex/20171016_POOL_POS_1_105-134.mzML"</span></span>
<span><span class="co">## [3] "/__w/_temp/Library/msdata/sciex/20171016_POOL_POS_1_105-134.mzML"</span></span>
<span><span class="co">## [4] "/__w/_temp/Library/msdata/sciex/20171016_POOL_POS_1_105-134.mzML"</span></span>
<span><span class="co">## [5] "/__w/_temp/Library/msdata/sciex/20171016_POOL_POS_1_105-134.mzML"</span></span>
<span><span class="co">## [6] "/__w/_temp/Library/msdata/sciex/20171016_POOL_POS_1_105-134.mzML"</span></span></code></pre>
<p>The figure below illustrates this link. With that last call we have
thus established links between samples and 3 different data elements in
the <code>MsExperiment</code>.</p>
<p><img src="imgs/Links_03.png"><!-- --></p>
<div class="sourceCode" id="cb69"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">lmse</span></span></code></pre></div>
<pre><code><span><span class="co">## Object of class MsExperiment </span></span>
<span><span class="co">##  Files: mzML_files, annotations </span></span>
<span><span class="co">##  Spectra: MS1 (1862) </span></span>
<span><span class="co">##  Experiment data: 2 sample(s)</span></span>
<span><span class="co">##  Sample data links:</span></span>
<span><span class="co">##   - experimentFiles.mzML_file: 2 sample(s) to 2 element(s).</span></span>
<span><span class="co">##   - experimentFiles.annotations: 2 sample(s) to 1 element(s).</span></span>
<span><span class="co">##   - spectra: 2 sample(s) to 1862 element(s).</span></span></code></pre>
<p>A convenience function to quickly extract the index of a sample a
spectrum is associated with is <code><a href="../reference/MsExperiment.html">spectraSampleIndex()</a></code>. This
function returns an <code>integer</code> vector of length equal to the
number of spectra in the object with the row in the object’s
<code>sampleData</code> a spectrum is linked to, or
<code>NA_integer_</code> if a spectrum is not linked to any sample.</p>
<div class="sourceCode" id="cb71"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co">#' Show the sample assignment for the first few spectra</span></span>
<span><span class="fu"><a href="../reference/MsExperiment.html">spectraSampleIndex</a></span><span class="op">(</span><span class="va">lmse</span><span class="op">)</span> <span class="op">|&gt;</span></span>
<span>    <span class="fu"><a href="https://rdrr.io/r/utils/head.html" class="external-link">head</a></span><span class="op">(</span><span class="op">)</span></span></code></pre></div>
<pre><code><span><span class="co">## [1] 1 1 1 1 1 1</span></span></code></pre>
<p>If we had also quantified <em>feature</em> values, we could also link
them to the samples. Below we create a simple, small
<code>SummarizedExperiment</code> to represent such quantified feature
values and add that to our experiment. To show that
<code>MsExperiment</code> supports also links between subsets of data
elements, we create a <code>SummarizedExperiment</code> that contains
values for an additional sample which is not present in our
<code>sampleData</code>. Also, we add samples in an arbitrary order.</p>
<div class="sourceCode" id="cb73"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va"><a href="https://bioconductor.org/packages/SummarizedExperiment" class="external-link">SummarizedExperiment</a></span><span class="op">)</span></span>
<span><span class="va">sd</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/S4Vectors/man/DataFrame-class.html" class="external-link">DataFrame</a></span><span class="op">(</span>sample <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"QC2"</span>, <span class="st">"QC1"</span>, <span class="st">"QC3"</span><span class="op">)</span>, idx <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="fl">3</span>, <span class="fl">1</span>, <span class="fl">5</span><span class="op">)</span><span class="op">)</span></span>
<span><span class="va">se</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/SummarizedExperiment/man/SummarizedExperiment-class.html" class="external-link">SummarizedExperiment</a></span><span class="op">(</span>colData <span class="op">=</span> <span class="va">sd</span>, assay <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/pkg/BiocGenerics/man/cbind.html" class="external-link">cbind</a></span><span class="op">(</span><span class="fl">1</span><span class="op">:</span><span class="fl">10</span>, <span class="fl">11</span><span class="op">:</span><span class="fl">20</span>, <span class="fl">21</span><span class="op">:</span><span class="fl">30</span><span class="op">)</span><span class="op">)</span></span>
<span></span>
<span><span class="fu"><a href="../reference/MsExperiment.html">qdata</a></span><span class="op">(</span><span class="va">lmse</span><span class="op">)</span> <span class="op">&lt;-</span> <span class="va">se</span></span></code></pre></div>
<p>Next we link the samples in this <code>SummarizedExperiment</code> to
the samples to in the <code>MsExperiment</code> using matching values
between the <code>"sample_id"</code> column in the object’s
<code>sampleData</code> data frame and the column <code>"sample"</code>
in the <code>SummarizedExperiment</code>’s <code>colData</code> which is
stored in the <code>@qdata</code> slot. The naming convention to define
such columns is <code>&lt;slot name&gt;.&lt;column name&gt;</code>.</p>
<div class="sourceCode" id="cb74"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="../reference/MsExperiment.html">sampleData</a></span><span class="op">(</span><span class="va">lmse</span><span class="op">)</span><span class="op">$</span><span class="va">sample_id</span></span></code></pre></div>
<pre><code><span><span class="co">## [1] "QC1" "QC2"</span></span></code></pre>
<div class="sourceCode" id="cb76"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="../reference/MsExperiment.html">qdata</a></span><span class="op">(</span><span class="va">lmse</span><span class="op">)</span><span class="op">$</span><span class="va">sample</span></span></code></pre></div>
<pre><code><span><span class="co">## [1] "QC2" "QC1" "QC3"</span></span></code></pre>
<div class="sourceCode" id="cb78"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">lmse</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/MsExperiment.html">linkSampleData</a></span><span class="op">(</span><span class="va">lmse</span>, with <span class="op">=</span> <span class="st">"sampleData.sample_id = qdata.sample"</span><span class="op">)</span></span>
<span><span class="va">lmse</span></span></code></pre></div>
<pre><code><span><span class="co">## Object of class MsExperiment </span></span>
<span><span class="co">##  Files: mzML_files, annotations </span></span>
<span><span class="co">##  Spectra: MS1 (1862) </span></span>
<span><span class="co">##  SummarizedExperiment: 10 feature(s)</span></span>
<span><span class="co">##  Experiment data: 2 sample(s)</span></span>
<span><span class="co">##  Sample data links:</span></span>
<span><span class="co">##   - experimentFiles.mzML_file: 2 sample(s) to 2 element(s).</span></span>
<span><span class="co">##   - experimentFiles.annotations: 2 sample(s) to 1 element(s).</span></span>
<span><span class="co">##   - spectra: 2 sample(s) to 1862 element(s).</span></span>
<span><span class="co">##   - qdata: 2 sample(s) to 2 column(s).</span></span></code></pre>
<p>The main advantage of all these links is that any subsetting of the
experiment by sample will keep the (linked) data consistent. To
illustrate this we subset below the experiment to the second sample.</p>
<div class="sourceCode" id="cb80"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">b</span> <span class="op">&lt;-</span> <span class="va">lmse</span><span class="op">[</span><span class="fl">2</span><span class="op">]</span></span>
<span><span class="va">b</span></span></code></pre></div>
<pre><code><span><span class="co">## Object of class MsExperiment </span></span>
<span><span class="co">##  Files: mzML_files, annotations, mzML_file </span></span>
<span><span class="co">##  Spectra: MS1 (931) </span></span>
<span><span class="co">##  SummarizedExperiment: 10 feature(s)</span></span>
<span><span class="co">##  Experiment data: 1 sample(s)</span></span>
<span><span class="co">##  Sample data links:</span></span>
<span><span class="co">##   - experimentFiles.mzML_file: 1 sample(s) to 1 element(s).</span></span>
<span><span class="co">##   - experimentFiles.annotations: 1 sample(s) to 1 element(s).</span></span>
<span><span class="co">##   - spectra: 1 sample(s) to 931 element(s).</span></span>
<span><span class="co">##   - qdata: 1 sample(s) to 1 column(s).</span></span></code></pre>
<p>The subset object contains now all data elements that are linked to
this second sample. Accessing the <code>assay</code> of the
<code>SummarizedExperiment</code> in <code>qdata</code> will thus return
only the quantified feature abundances for this second sample.</p>
<div class="sourceCode" id="cb82"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/pkg/SummarizedExperiment/man/SummarizedExperiment-class.html" class="external-link">assay</a></span><span class="op">(</span><span class="fu"><a href="../reference/MsExperiment.html">qdata</a></span><span class="op">(</span><span class="va">b</span><span class="op">)</span><span class="op">)</span></span></code></pre></div>
<pre><code><span><span class="co">##       [,1]</span></span>
<span><span class="co">##  [1,]    1</span></span>
<span><span class="co">##  [2,]    2</span></span>
<span><span class="co">##  [3,]    3</span></span>
<span><span class="co">##  [4,]    4</span></span>
<span><span class="co">##  [5,]    5</span></span>
<span><span class="co">##  [6,]    6</span></span>
<span><span class="co">##  [7,]    7</span></span>
<span><span class="co">##  [8,]    8</span></span>
<span><span class="co">##  [9,]    9</span></span>
<span><span class="co">## [10,]   10</span></span></code></pre>
<p>But what happens for data elements that are not linked to any sample?
Below we add a <code>data.frame</code> as a <code>metadata</code> to the
experiment and subset the object again.</p>
<div class="sourceCode" id="cb84"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/pkg/S4Vectors/man/Annotated-class.html" class="external-link">metadata</a></span><span class="op">(</span><span class="va">lmse</span><span class="op">)</span><span class="op">$</span><span class="va">other</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/data.frame.html" class="external-link">data.frame</a></span><span class="op">(</span></span>
<span>                  sample_name <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"study_1"</span>, <span class="st">"POOL"</span>, <span class="st">"study_2"</span><span class="op">)</span>,</span>
<span>                  index <span class="op">=</span> <span class="fl">1</span><span class="op">:</span><span class="fl">3</span><span class="op">)</span></span>
<span><span class="va">b</span> <span class="op">&lt;-</span> <span class="va">lmse</span><span class="op">[</span><span class="fl">2</span><span class="op">]</span></span>
<span><span class="fu"><a href="https://rdrr.io/pkg/S4Vectors/man/Annotated-class.html" class="external-link">metadata</a></span><span class="op">(</span><span class="va">b</span><span class="op">)</span></span></code></pre></div>
<pre><code><span><span class="co">## $other</span></span>
<span><span class="co">##   sample_name index</span></span>
<span><span class="co">## 1     study_1     1</span></span>
<span><span class="co">## 2        POOL     2</span></span>
<span><span class="co">## 3     study_2     3</span></span></code></pre>
<p>By default, any element which is <strong>not</strong> linked to a
sample is retained in the filtered/subset object.</p>
<p>We next link each sample to the second row in this data frame and
subset the data again to the second sample.</p>
<div class="sourceCode" id="cb86"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">lmse</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/MsExperiment.html">linkSampleData</a></span><span class="op">(</span><span class="va">lmse</span>, with <span class="op">=</span> <span class="st">"metadata.other"</span>,</span>
<span>                      sampleIndex <span class="op">=</span> <span class="fl">1</span><span class="op">:</span><span class="fl">2</span>, withIndex <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="fl">2</span>, <span class="fl">2</span><span class="op">)</span><span class="op">)</span></span>
<span><span class="va">b</span> <span class="op">&lt;-</span> <span class="va">lmse</span><span class="op">[</span><span class="fl">2</span><span class="op">]</span></span>
<span><span class="fu"><a href="https://rdrr.io/pkg/S4Vectors/man/Annotated-class.html" class="external-link">metadata</a></span><span class="op">(</span><span class="va">b</span><span class="op">)</span></span></code></pre></div>
<pre><code><span><span class="co">## $other</span></span>
<span><span class="co">##   sample_name index</span></span>
<span><span class="co">## 2        POOL     2</span></span></code></pre>
<p>Subsetting thus retained only the row in the data frame for the
linked sample. Obviously it is also possible to subset to multiple
samples, in arbitrary order. Below we re-order our experiment.</p>
<div class="sourceCode" id="cb88"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">lmse</span> <span class="op">&lt;-</span> <span class="va">lmse</span><span class="op">[</span><span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="fl">2</span>, <span class="fl">1</span><span class="op">)</span><span class="op">]</span></span>
<span><span class="fu"><a href="../reference/MsExperiment.html">sampleData</a></span><span class="op">(</span><span class="va">lmse</span><span class="op">)</span></span></code></pre></div>
<pre><code><span><span class="co">## DataFrame with 2 rows and 4 columns</span></span>
<span><span class="co">##     sample_id sample_name injection_idx      raw_file</span></span>
<span><span class="co">##   &lt;character&gt; &lt;character&gt;     &lt;numeric&gt;   &lt;character&gt;</span></span>
<span><span class="co">## 1         QC2     QC Pool             3 /__w/_temp...</span></span>
<span><span class="co">## 2         QC1     QC Pool             1 /__w/_temp...</span></span></code></pre>
<p>The sample order is thus reversed and also all other linked elements
are re-ordered accordingly, such as <code>"mzML_file"</code> in the
object’s <code>experimentFiles</code>.</p>
<div class="sourceCode" id="cb90"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="../reference/MsExperiment.html">experimentFiles</a></span><span class="op">(</span><span class="va">lmse</span><span class="op">)</span><span class="op">$</span><span class="va">mzML_file</span></span></code></pre></div>
<pre><code><span><span class="co">## [1] "/__w/_temp/Library/msdata/sciex/20171016_POOL_POS_3_105-134.mzML"</span></span>
<span><span class="co">## [2] "/__w/_temp/Library/msdata/sciex/20171016_POOL_POS_1_105-134.mzML"</span></span></code></pre>
<p>It is however important to note, that subsetting will also
<em>duplicate</em> elements that are associated with multiple
samples:</p>
<div class="sourceCode" id="cb92"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="../reference/MsExperiment.html">experimentFiles</a></span><span class="op">(</span><span class="va">lmse</span><span class="op">)</span><span class="op">$</span><span class="va">annotations</span></span></code></pre></div>
<pre><code><span><span class="co">## [1] "internal_standards.txt" "internal_standards.txt"</span></span></code></pre>
<p>Thus, while we added a single <em>annotation</em> file to the data
element <code>"annotations"</code> in <code>experimentFiles</code>,
after subsetting we ended up with two identical files. This duplication
of <em>n:m</em> relationships between samples to elements does however
not affect data consistency. A sample will always be linked to the
correct value/element.</p>
</div>
<div class="section level2">
<h2 id="subset-and-filter-msexperiment">Subset and filter <code>MsExperiment</code><a class="anchor" aria-label="anchor" href="#subset-and-filter-msexperiment"></a>
</h2>
<p>As already shown above, <code>MsExperiment</code> objects can be
subset with <code>[</code> which will subset the data by sample.
Depending on whether relationships (links) between samples and any other
data within the object are present also these are correctly subset. In
addition to this general subset operation, it is possible to
individually filter the spectra data within an <code>MsExperiment</code>
using the <code>filterSpectra</code> function. This function takes any
filter function supported by <code>Spectra</code> with parameter
<code>filter</code>. Parameters for this filter function can be passed
through <code>...</code>. As an example we filter below the spectra data
of our <code>MsExperiment</code> keeping only spectra with an retention
time between 200 and 210 seconds.</p>
<div class="sourceCode" id="cb94"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co">#' Filter the Spectra using the `filterRt` function providing also the</span></span>
<span><span class="co">#' parameters for this function.</span></span>
<span><span class="va">res</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/ProtGenerics/man/filterSpectra.html" class="external-link">filterSpectra</a></span><span class="op">(</span><span class="va">lmse</span>, <span class="va">filterRt</span>, rt <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="fl">200</span>, <span class="fl">210</span><span class="op">)</span><span class="op">)</span></span>
<span><span class="va">res</span></span></code></pre></div>
<pre><code><span><span class="co">## Object of class MsExperiment </span></span>
<span><span class="co">##  Files: mzML_files, annotations, mzML_file </span></span>
<span><span class="co">##  Spectra: MS1 (72) </span></span>
<span><span class="co">##  SummarizedExperiment: 10 feature(s)</span></span>
<span><span class="co">##  Experiment data: 2 sample(s)</span></span>
<span><span class="co">##  Sample data links:</span></span>
<span><span class="co">##   - experimentFiles.mzML_file: 2 sample(s) to 2 element(s).</span></span>
<span><span class="co">##   - experimentFiles.annotations: 2 sample(s) to 2 element(s).</span></span>
<span><span class="co">##   - spectra: 2 sample(s) to 72 element(s).</span></span>
<span><span class="co">##   - qdata: 2 sample(s) to 2 column(s).</span></span>
<span><span class="co">##   - metadata.other: 2 sample(s) to 2 element(s).</span></span></code></pre>
<p>The resulting <code>MsExperiment</code> contains now much fewer
spectra. <code>filterSpectra</code> did only filter the spectra data,
but not any of the other data slots. It did however update and
consolidate the relationships between samples and spectra (if present)
after filtering:</p>
<div class="sourceCode" id="cb96"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co">#' Extract spectra of the second sample after filtering</span></span>
<span><span class="fu"><a href="https://rdrr.io/pkg/ProtGenerics/man/protgenerics.html" class="external-link">spectra</a></span><span class="op">(</span><span class="va">res</span><span class="op">[</span><span class="fl">2L</span><span class="op">]</span><span class="op">)</span></span></code></pre></div>
<pre><code><span><span class="co">## MSn data (Spectra) with 36 spectra in a MsBackendMzR backend:</span></span>
<span><span class="co">##       msLevel     rtime scanIndex</span></span>
<span><span class="co">##     &lt;integer&gt; &lt;numeric&gt; &lt;integer&gt;</span></span>
<span><span class="co">## 1           1   200.049       717</span></span>
<span><span class="co">## 2           1   200.328       718</span></span>
<span><span class="co">## 3           1   200.607       719</span></span>
<span><span class="co">## 4           1   200.886       720</span></span>
<span><span class="co">## 5           1   201.165       721</span></span>
<span><span class="co">## ...       ...       ...       ...</span></span>
<span><span class="co">## 32          1   208.698       748</span></span>
<span><span class="co">## 33          1   208.977       749</span></span>
<span><span class="co">## 34          1   209.256       750</span></span>
<span><span class="co">## 35          1   209.535       751</span></span>
<span><span class="co">## 36          1   209.814       752</span></span>
<span><span class="co">##  ... 34 more variables/columns.</span></span>
<span><span class="co">## </span></span>
<span><span class="co">## file(s):</span></span>
<span><span class="co">## 20171016_POOL_POS_1_105-134.mzML</span></span>
<span><span class="co">## Processing:</span></span>
<span><span class="co">##  Filter: select retention time [200..210] on MS level(s)  [Wed Jun  4 06:39:56 2025]</span></span></code></pre>
</div>
<div class="section level2">
<h2 id="using-msexperiment-with-msbackendsql">Using <code>MsExperiment</code> with <code>MsBackendSql</code><a class="anchor" aria-label="anchor" href="#using-msexperiment-with-msbackendsql"></a>
</h2>
<p>The <em><a href="https://bioconductor.org/packages/3.22/MsBackendSql" class="external-link">MsBackendSql</a></em>
provides functionality to store mass spectrometry data in a SQL database
and to <em>represent</em> such data as a <code>Spectra</code> object
(through its <code>MsBackendSql</code> or
<code>MsBackendOfflineSql</code> backends). Next to the MS data, it is
also possible to store sample data in such SQL databases hence allowing
to create self-contained databases for MS experiments. This simplifies
the process to load pre-defined MS experiment or to share experiment
data across analyses or with a collaborator. In this section we show how
MS data and sample information from an experiment can be stored into a
SQL database and how that data can then be loaded again as a
<code>MsExperiment</code>.</p>
<p>Below we first define the raw data files of the experiment. In our
example we use two small test files provided with the <em><a href="https://bioconductor.org/packages/3.22/msdata" class="external-link">msdata</a></em>
package.</p>
<div class="sourceCode" id="cb98"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">fls</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/system.file.html" class="external-link">system.file</a></span><span class="op">(</span><span class="st">"microtofq"</span>, <span class="st">"MM14.mzML"</span>, package <span class="op">=</span> <span class="st">"msdata"</span><span class="op">)</span>,</span>
<span>         <span class="fu"><a href="https://rdrr.io/r/base/system.file.html" class="external-link">system.file</a></span><span class="op">(</span><span class="st">"microtofq"</span>, <span class="st">"MM8.mzML"</span>, package <span class="op">=</span> <span class="st">"msdata"</span><span class="op">)</span><span class="op">)</span></span></code></pre></div>
<p>We can then use the <code>createMsBackendSqlDatabase</code> function
from the <em><a href="https://bioconductor.org/packages/3.22/MsBackendSql" class="external-link">MsBackendSql</a></em>
package to write the data into a SQLite database. SQLite databases are
particularly easy to create, use and also to share, but for large
experiments it is suggested to use more powerful database engines, such
as for example <em>MariaDB</em> (see also the documentation and vignette
of the <em><a href="https://bioconductor.org/packages/3.22/MsBackendSql" class="external-link">MsBackendSql</a></em>
package for parameters and settings for such large databases).</p>
<p>Also, in our example we store the SQLite database into a data
temporary file, but for a real use case we would want to use a
<em>real</em> file.</p>
<div class="sourceCode" id="cb99"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va"><a href="https://github.com/RforMassSpectrometry/MsBackendSql" class="external-link">MsBackendSql</a></span><span class="op">)</span></span>
<span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va"><a href="https://rsqlite.r-dbi.org" class="external-link">RSQLite</a></span><span class="op">)</span></span>
<span></span>
<span><span class="co">#' Create the SQLite database where to store the data. For our</span></span>
<span><span class="co">#' example we create the database in a temporary file.</span></span>
<span><span class="va">sql_file</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/tempfile.html" class="external-link">tempfile</a></span><span class="op">(</span><span class="op">)</span></span>
<span><span class="va">con</span> <span class="op">&lt;-</span> <span class="fu">dbConnect</span><span class="op">(</span><span class="fu"><a href="https://rsqlite.r-dbi.org/reference/SQLite.html" class="external-link">SQLite</a></span><span class="op">(</span><span class="op">)</span>, dbname <span class="op">=</span> <span class="va">sql_file</span><span class="op">)</span></span>
<span></span>
<span><span class="co">#' Write the MS data to the.</span></span>
<span><span class="fu"><a href="https://rdrr.io/pkg/MsBackendSql/man/MsBackendSql.html" class="external-link">createMsBackendSqlDatabase</a></span><span class="op">(</span>dbcon <span class="op">=</span> <span class="va">con</span>, <span class="va">fls</span><span class="op">)</span></span></code></pre></div>
<pre><code><span><span class="co">## Importing data ...</span></span></code></pre>
<pre><code><span><span class="co">## [==========================================================] 1/1 (100%) in  1s</span></span>
<span><span class="co">## Creating indices .... Done</span></span></code></pre>
<pre><code><span><span class="co">## [1] TRUE</span></span></code></pre>
<p>We can now create a <code>Spectra</code> object representing the MS
data in that database using either a <code>MsBackendSql</code> or
<code>MsBackendOfflineSql</code> backend.</p>
<div class="sourceCode" id="cb103"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co">#' Load the database as a Spectra object</span></span>
<span><span class="va">sps</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/Spectra/man/Spectra.html" class="external-link">Spectra</a></span><span class="op">(</span><span class="va">sql_file</span>, source <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/pkg/MsBackendSql/man/MsBackendOfflineSql.html" class="external-link">MsBackendOfflineSql</a></span><span class="op">(</span><span class="op">)</span>, drv <span class="op">=</span> <span class="fu"><a href="https://rsqlite.r-dbi.org/reference/SQLite.html" class="external-link">SQLite</a></span><span class="op">(</span><span class="op">)</span><span class="op">)</span></span>
<span><span class="va">sps</span></span></code></pre></div>
<pre><code><span><span class="co">## MSn data (Spectra) with 310 spectra in a MsBackendOfflineSql backend:</span></span>
<span><span class="co">##       msLevel precursorMz  polarity</span></span>
<span><span class="co">##     &lt;integer&gt;   &lt;numeric&gt; &lt;integer&gt;</span></span>
<span><span class="co">## 1           1          NA         1</span></span>
<span><span class="co">## 2           1          NA         1</span></span>
<span><span class="co">## 3           1          NA         1</span></span>
<span><span class="co">## 4           1          NA         1</span></span>
<span><span class="co">## 5           1          NA         1</span></span>
<span><span class="co">## ...       ...         ...       ...</span></span>
<span><span class="co">## 306         1          NA         1</span></span>
<span><span class="co">## 307         1          NA         1</span></span>
<span><span class="co">## 308         1          NA         1</span></span>
<span><span class="co">## 309         1          NA         1</span></span>
<span><span class="co">## 310         1          NA         1</span></span>
<span><span class="co">##  ... 35 more variables/columns.</span></span>
<span><span class="co">##  Use  'spectraVariables' to list all of them.</span></span>
<span><span class="co">## Database: /tmp/RtmpHXW8vs/file146c7b850b36</span></span></code></pre>
<p>Next we define sample annotations for the two files and create a
<code>MsExperiment</code> with the MS data and the related sample
annotations.</p>
<div class="sourceCode" id="cb105"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co">#' Define sample annotations. Ideally all experiment-relavant information</span></span>
<span><span class="co">#' should be specified.</span></span>
<span><span class="va">sdata</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/data.frame.html" class="external-link">data.frame</a></span><span class="op">(</span>file_name <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/pkg/BiocGenerics/man/path.html" class="external-link">basename</a></span><span class="op">(</span><span class="va">fls</span><span class="op">)</span>,</span>
<span>                    sample_name <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"MM14"</span>, <span class="st">"MM8"</span><span class="op">)</span>,</span>
<span>                    sample_group <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"ctrl"</span>, <span class="st">"ctrl"</span><span class="op">)</span><span class="op">)</span></span>
<span></span>
<span><span class="co">#' To simplify subsequent linking between samples and data files we</span></span>
<span><span class="co">#' define a new spectra variable with the file names of the raw data files.</span></span>
<span><span class="va">sps</span><span class="op">$</span><span class="va">file_name</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/BiocGenerics/man/path.html" class="external-link">basename</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/pkg/ProtGenerics/man/protgenerics.html" class="external-link">dataOrigin</a></span><span class="op">(</span><span class="va">sps</span><span class="op">)</span><span class="op">)</span></span>
<span></span>
<span><span class="co">#' Create the MsExperiment for that data</span></span>
<span><span class="va">mse</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/MsExperiment.html">MsExperiment</a></span><span class="op">(</span>sampleData <span class="op">=</span> <span class="va">sdata</span>, spectra <span class="op">=</span> <span class="va">sps</span><span class="op">)</span></span></code></pre></div>
<p>Finally we <em>link</em> the MS spectra to the individual samples
using the file names of the original data files.</p>
<div class="sourceCode" id="cb106"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co">#' Establish the mapping between spectra and samples</span></span>
<span><span class="va">mse</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/MsExperiment.html">linkSampleData</a></span><span class="op">(</span><span class="va">mse</span>, with <span class="op">=</span> <span class="st">"sampleData.file_name = spectra.file_name"</span><span class="op">)</span></span>
<span><span class="va">mse</span></span></code></pre></div>
<pre><code><span><span class="co">## Object of class MsExperiment </span></span>
<span><span class="co">##  Spectra: MS1 (310) </span></span>
<span><span class="co">##  Experiment data: 2 sample(s)</span></span>
<span><span class="co">##  Sample data links:</span></span>
<span><span class="co">##   - spectra: 2 sample(s) to 310 element(s).</span></span></code></pre>
<p>With sample annotations and the mapping between samples and spectra
defined, we can next write this information to the SQL database
containing already the MS data of our experiment: the
<code>dbWriteSampleData</code> function writes, for
<code>MsExperiment</code> objects that use a <code>Spectra</code> with a
<code>MsBackendSql</code> (or <code>MsBackendOfflineSql</code>) backend
the available sample data to additional database tables in the
database.</p>
<div class="sourceCode" id="cb108"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co">#' Write sample data (and mapping between samples and spectra) to the</span></span>
<span><span class="co">#' database</span></span>
<span><span class="fu"><a href="../reference/dbWriteSampleData.html">dbWriteSampleData</a></span><span class="op">(</span><span class="va">mse</span><span class="op">)</span></span></code></pre></div>
<p>The SQL database does now contain the MS data, the sample annotations
and the mapping between samples and spectra. Creation of such a database
needs to be performed only once for an experiment. The associated SQLite
database file could now be shared with a collaborator or used across
different analyses without the need to share or copy also the raw data
files or sample annotations (which would usually needed to be provided
as a separate file).</p>
<p>To load the MS experiment data again into R, we first need to create
a <code>Spectra</code> object with a <code>MsBackendSql</code> or
<code>MsBackendOfflineSql</code> backend for that SQL database.</p>
<div class="sourceCode" id="cb109"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co">#' Create a Spectra object for the MS data of that database</span></span>
<span><span class="va">sps</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/Spectra/man/Spectra.html" class="external-link">Spectra</a></span><span class="op">(</span><span class="va">sql_file</span>, source <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/pkg/MsBackendSql/man/MsBackendOfflineSql.html" class="external-link">MsBackendOfflineSql</a></span><span class="op">(</span><span class="op">)</span>, drv <span class="op">=</span> <span class="fu"><a href="https://rsqlite.r-dbi.org/reference/SQLite.html" class="external-link">SQLite</a></span><span class="op">(</span><span class="op">)</span><span class="op">)</span></span></code></pre></div>
<p>We then pass this <code>Spectra</code> object with parameter
<code>spectra</code> to the <code>MsExperiment</code> constructor call.
The function will then read also sample data from the database (if
available) as well as the mapping between samples and spectra to restore
our MS experiment.</p>
<div class="sourceCode" id="cb110"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co">#' Create the MsExperiment reading MS data and sample data from</span></span>
<span><span class="co">#' the database</span></span>
<span><span class="va">mse</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/MsExperiment.html">MsExperiment</a></span><span class="op">(</span>spectra <span class="op">=</span> <span class="va">sps</span><span class="op">)</span></span>
<span><span class="va">mse</span></span></code></pre></div>
<pre><code><span><span class="co">## Object of class MsExperiment </span></span>
<span><span class="co">##  Spectra: MS1 (310) </span></span>
<span><span class="co">##  Experiment data: 2 sample(s)</span></span>
<span><span class="co">##  Sample data links:</span></span>
<span><span class="co">##   - spectra: 2 sample(s) to 310 element(s).</span></span></code></pre>
<p>We have thus now the full MS experiment data available.</p>
<p>For smaller experiments, such SQL databases could also be created
with an alternative, eventually simpler, approach. For that we first use
the <code>readMsExperiment</code> function to create an
<code>MsExperiment</code> from the file names of the raw MS data files
and a <code>data.frame</code> with the associated sample annotations. We
re-use here the variables <code>fls</code> with the data file names and
<code>sdata</code> with the sample data.</p>
<div class="sourceCode" id="cb112"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co">#' Create an MsExperiment with data from two (raw) data files</span></span>
<span><span class="co">#' and associated sample annotations.</span></span>
<span><span class="va">mse</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/readMsExperiment.html">readMsExperiment</a></span><span class="op">(</span>spectraFiles <span class="op">=</span> <span class="va">fls</span>, sampleData <span class="op">=</span> <span class="va">sdata</span><span class="op">)</span></span>
<span><span class="va">mse</span></span></code></pre></div>
<pre><code><span><span class="co">## Object of class MsExperiment </span></span>
<span><span class="co">##  Spectra: MS1 (310) </span></span>
<span><span class="co">##  Experiment data: 2 sample(s)</span></span>
<span><span class="co">##  Sample data links:</span></span>
<span><span class="co">##   - spectra: 2 sample(s) to 310 element(s).</span></span></code></pre>
<p>The MS data in that <code>MsExperiment</code> is represented by a
<code>Spectra</code> object using the <code>MsBackendMzR</code>:</p>
<div class="sourceCode" id="cb114"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co">#' Show the Spectra from the experiment</span></span>
<span><span class="fu"><a href="https://rdrr.io/pkg/ProtGenerics/man/protgenerics.html" class="external-link">spectra</a></span><span class="op">(</span><span class="va">mse</span><span class="op">)</span></span></code></pre></div>
<pre><code><span><span class="co">## MSn data (Spectra) with 310 spectra in a MsBackendMzR backend:</span></span>
<span><span class="co">##       msLevel     rtime scanIndex</span></span>
<span><span class="co">##     &lt;integer&gt; &lt;numeric&gt; &lt;integer&gt;</span></span>
<span><span class="co">## 1           1   270.334         1</span></span>
<span><span class="co">## 2           1   270.671         2</span></span>
<span><span class="co">## 3           1   271.007         3</span></span>
<span><span class="co">## 4           1   271.343         4</span></span>
<span><span class="co">## 5           1   271.680         5</span></span>
<span><span class="co">## ...       ...       ...       ...</span></span>
<span><span class="co">## 306         1   65.4360       194</span></span>
<span><span class="co">## 307         1   65.7720       195</span></span>
<span><span class="co">## 308         1   66.1092       196</span></span>
<span><span class="co">## 309         1   66.4458       197</span></span>
<span><span class="co">## 310         1   66.7818       198</span></span>
<span><span class="co">##  ... 34 more variables/columns.</span></span>
<span><span class="co">## </span></span>
<span><span class="co">## file(s):</span></span>
<span><span class="co">## MM14.mzML</span></span>
<span><span class="co">## MM8.mzML</span></span></code></pre>
<p>We next change the backend of this <code>Spectra</code> object using
the <code>setBackend</code> function to a
<code>MsBackendOfflineSql</code>. We choose again a SQLite database
format and store the data to a temporary file.</p>
<div class="sourceCode" id="cb116"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co">#' Define the SQLite database file</span></span>
<span><span class="va">sql_file2</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/tempfile.html" class="external-link">tempfile</a></span><span class="op">(</span><span class="op">)</span></span>
<span></span>
<span><span class="co">#' Change the backend of the Spectra data</span></span>
<span><span class="fu"><a href="https://rdrr.io/pkg/ProtGenerics/man/protgenerics.html" class="external-link">spectra</a></span><span class="op">(</span><span class="va">mse</span><span class="op">)</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/ProtGenerics/man/backendInitialize.html" class="external-link">setBackend</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/pkg/ProtGenerics/man/protgenerics.html" class="external-link">spectra</a></span><span class="op">(</span><span class="va">mse</span><span class="op">)</span>, <span class="fu"><a href="https://rdrr.io/pkg/MsBackendSql/man/MsBackendOfflineSql.html" class="external-link">MsBackendOfflineSql</a></span><span class="op">(</span><span class="op">)</span>,</span>
<span>                           dbname <span class="op">=</span> <span class="va">sql_file2</span>, drv <span class="op">=</span> <span class="fu"><a href="https://rsqlite.r-dbi.org/reference/SQLite.html" class="external-link">SQLite</a></span><span class="op">(</span><span class="op">)</span><span class="op">)</span></span></code></pre></div>
<pre><code><span><span class="co">## Importing data ...</span></span></code></pre>
<pre><code><span><span class="co">## [==========================================================] 2/2 (100%) in  0s</span></span>
<span><span class="co">## Creating indices .... Done</span></span></code></pre>
<div class="sourceCode" id="cb119"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/pkg/ProtGenerics/man/protgenerics.html" class="external-link">spectra</a></span><span class="op">(</span><span class="va">mse</span><span class="op">)</span></span></code></pre></div>
<pre><code><span><span class="co">## MSn data (Spectra) with 310 spectra in a MsBackendOfflineSql backend:</span></span>
<span><span class="co">##       msLevel precursorMz  polarity</span></span>
<span><span class="co">##     &lt;integer&gt;   &lt;numeric&gt; &lt;integer&gt;</span></span>
<span><span class="co">## 1           1          NA         1</span></span>
<span><span class="co">## 2           1          NA         1</span></span>
<span><span class="co">## 3           1          NA         1</span></span>
<span><span class="co">## 4           1          NA         1</span></span>
<span><span class="co">## 5           1          NA         1</span></span>
<span><span class="co">## ...       ...         ...       ...</span></span>
<span><span class="co">## 306         1          NA         1</span></span>
<span><span class="co">## 307         1          NA         1</span></span>
<span><span class="co">## 308         1          NA         1</span></span>
<span><span class="co">## 309         1          NA         1</span></span>
<span><span class="co">## 310         1          NA         1</span></span>
<span><span class="co">##  ... 35 more variables/columns.</span></span>
<span><span class="co">##  Use  'spectraVariables' to list all of them.</span></span>
<span><span class="co">## Database: /tmp/RtmpHXW8vs/file146c5f0aa01a</span></span>
<span><span class="co">## Processing:</span></span>
<span><span class="co">##  Switch backend from MsBackendMzR to MsBackendOfflineSql [Wed Jun  4 06:39:59 2025]</span></span></code></pre>
<p>All MS data is now stored in the SQLite database and the
<code>Spectra</code> object of our experiment uses the
<code>MsBackendOfflineSql</code> to represent/interface this data. To
store also the sample data to the database we can use the
<code>dbWriteSampleData</code> as before.</p>
<div class="sourceCode" id="cb121"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co">#' Store also sample data to the database</span></span>
<span><span class="fu"><a href="../reference/dbWriteSampleData.html">dbWriteSampleData</a></span><span class="op">(</span><span class="va">mse</span><span class="op">)</span></span></code></pre></div>
<p>In addition to create self-contained SQL databases for an experiment,
this code could also be used within an analysis workflow to
serialize/de-serialize results, always with the advantage of keeping the
full experiment data in a single place (file) in a language-agnostic
format.</p>
<p>We can for example use basic functions from the
<em>DBI</em>/<em>RSQLite</em> packages (or simple SQL commands) to
access the data in the database. Below we connect to the database and
list the available tables.</p>
<div class="sourceCode" id="cb122"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co">#' Connect to the database</span></span>
<span><span class="va">con</span> <span class="op">&lt;-</span> <span class="fu">dbConnect</span><span class="op">(</span><span class="fu"><a href="https://rsqlite.r-dbi.org/reference/SQLite.html" class="external-link">SQLite</a></span><span class="op">(</span><span class="op">)</span>, dbname <span class="op">=</span> <span class="va">sql_file2</span><span class="op">)</span></span>
<span></span>
<span><span class="co">#' List the available database tables</span></span>
<span><span class="fu">dbListTables</span><span class="op">(</span><span class="va">con</span><span class="op">)</span></span></code></pre></div>
<pre><code><span><span class="co">## [1] "msms_spectrum"            "msms_spectrum_peak_blob2"</span></span>
<span><span class="co">## [3] "sample_data"              "sample_to_msms_spectrum"</span></span></code></pre>
<p>The <em>msms_spectrum</em> table contains general metadata for the
individual MS spectra, <em>msms_spectrum_peak_blob</em> the actual peak
data (<em>m/z</em> and intensity values), <em>sample_data</em> the
sample annotations and <em>sample_to_msms_spectrum</em> the link between
samples and spectra. Below we retrieve the content of the
<em>sample_data</em> table.</p>
<div class="sourceCode" id="cb124"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co">#' Get the content of the sample_data table</span></span>
<span><span class="fu">dbGetQuery</span><span class="op">(</span><span class="va">con</span>, <span class="st">"select * from sample_data"</span><span class="op">)</span></span></code></pre></div>
<pre><code><span><span class="co">##   file_name sample_name sample_group</span></span>
<span><span class="co">## 1 MM14.mzML        MM14         ctrl</span></span>
<span><span class="co">## 2  MM8.mzML         MM8         ctrl</span></span>
<span><span class="co">##                                   spectraOrigin sample_id_</span></span>
<span><span class="co">## 1 /__w/_temp/Library/msdata/microtofq/MM14.mzML          1</span></span>
<span><span class="co">## 2  /__w/_temp/Library/msdata/microtofq/MM8.mzML          2</span></span></code></pre>
<div class="sourceCode" id="cb126"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co">#' Close the connection to the database</span></span>
<span><span class="fu">dbDisconnect</span><span class="op">(</span><span class="va">con</span><span class="op">)</span></span></code></pre></div>
</div>
<div class="section level2">
<h2 id="session-information">Session information<a class="anchor" aria-label="anchor" href="#session-information"></a>
</h2>
<div class="sourceCode" id="cb127"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/r/utils/sessionInfo.html" class="external-link">sessionInfo</a></span><span class="op">(</span><span class="op">)</span></span></code></pre></div>
<pre><code><span><span class="co">## R version 4.5.0 (2025-04-11)</span></span>
<span><span class="co">## Platform: x86_64-pc-linux-gnu</span></span>
<span><span class="co">## Running under: Ubuntu 24.04.2 LTS</span></span>
<span><span class="co">## </span></span>
<span><span class="co">## Matrix products: default</span></span>
<span><span class="co">## BLAS:   /usr/lib/x86_64-linux-gnu/openblas-pthread/libblas.so.3 </span></span>
<span><span class="co">## LAPACK: /usr/lib/x86_64-linux-gnu/openblas-pthread/libopenblasp-r0.3.26.so;  LAPACK version 3.12.0</span></span>
<span><span class="co">## </span></span>
<span><span class="co">## locale:</span></span>
<span><span class="co">##  [1] LC_CTYPE=en_US.UTF-8       LC_NUMERIC=C              </span></span>
<span><span class="co">##  [3] LC_TIME=en_US.UTF-8        LC_COLLATE=en_US.UTF-8    </span></span>
<span><span class="co">##  [5] LC_MONETARY=en_US.UTF-8    LC_MESSAGES=en_US.UTF-8   </span></span>
<span><span class="co">##  [7] LC_PAPER=en_US.UTF-8       LC_NAME=C                 </span></span>
<span><span class="co">##  [9] LC_ADDRESS=C               LC_TELEPHONE=C            </span></span>
<span><span class="co">## [11] LC_MEASUREMENT=en_US.UTF-8 LC_IDENTIFICATION=C       </span></span>
<span><span class="co">## </span></span>
<span><span class="co">## time zone: UTC</span></span>
<span><span class="co">## tzcode source: system (glibc)</span></span>
<span><span class="co">## </span></span>
<span><span class="co">## attached base packages:</span></span>
<span><span class="co">## [1] stats4    stats     graphics  grDevices utils     datasets  methods  </span></span>
<span><span class="co">## [8] base     </span></span>
<span><span class="co">## </span></span>
<span><span class="co">## other attached packages:</span></span>
<span><span class="co">##  [1] RSQLite_2.4.0               MsBackendSql_1.9.0         </span></span>
<span><span class="co">##  [3] SummarizedExperiment_1.39.0 Biobase_2.69.0             </span></span>
<span><span class="co">##  [5] GenomicRanges_1.61.0        GenomeInfoDb_1.45.4        </span></span>
<span><span class="co">##  [7] IRanges_2.43.0              MatrixGenerics_1.21.0      </span></span>
<span><span class="co">##  [9] matrixStats_1.5.0           rpx_2.15.1                 </span></span>
<span><span class="co">## [11] Spectra_1.19.2              BiocParallel_1.43.3        </span></span>
<span><span class="co">## [13] S4Vectors_0.47.0            BiocGenerics_0.55.0        </span></span>
<span><span class="co">## [15] generics_0.1.4              MsExperiment_1.11.1        </span></span>
<span><span class="co">## [17] ProtGenerics_1.39.2         BiocStyle_2.37.0           </span></span>
<span><span class="co">## </span></span>
<span><span class="co">## loaded via a namespace (and not attached):</span></span>
<span><span class="co">##  [1] DBI_1.2.3                   bitops_1.0-9               </span></span>
<span><span class="co">##  [3] httr2_1.1.2                 rlang_1.1.6                </span></span>
<span><span class="co">##  [5] magrittr_2.0.3              clue_0.3-66                </span></span>
<span><span class="co">##  [7] compiler_4.5.0              systemfonts_1.2.3          </span></span>
<span><span class="co">##  [9] vctrs_0.6.5                 reshape2_1.4.4             </span></span>
<span><span class="co">## [11] stringr_1.5.1               pkgconfig_2.0.3            </span></span>
<span><span class="co">## [13] MetaboCoreUtils_1.17.1      crayon_1.5.3               </span></span>
<span><span class="co">## [15] fastmap_1.2.0               dbplyr_2.5.0               </span></span>
<span><span class="co">## [17] XVector_0.49.0              rmarkdown_2.29             </span></span>
<span><span class="co">## [19] UCSC.utils_1.5.0            ragg_1.4.0                 </span></span>
<span><span class="co">## [21] purrr_1.0.4                 bit_4.6.0                  </span></span>
<span><span class="co">## [23] xfun_0.52                   MultiAssayExperiment_1.35.3</span></span>
<span><span class="co">## [25] cachem_1.1.0                jsonlite_2.0.0             </span></span>
<span><span class="co">## [27] progress_1.2.3              blob_1.2.4                 </span></span>
<span><span class="co">## [29] DelayedArray_0.35.1         prettyunits_1.2.0          </span></span>
<span><span class="co">## [31] parallel_4.5.0              cluster_2.1.8.1            </span></span>
<span><span class="co">## [33] R6_2.6.1                    bslib_0.9.0                </span></span>
<span><span class="co">## [35] stringi_1.8.7               jquerylib_0.1.4            </span></span>
<span><span class="co">## [37] Rcpp_1.0.14                 bookdown_0.43              </span></span>
<span><span class="co">## [39] knitr_1.50                  BiocBaseUtils_1.11.0       </span></span>
<span><span class="co">## [41] Matrix_1.7-3                igraph_2.1.4               </span></span>
<span><span class="co">## [43] tidyselect_1.2.1            abind_1.4-8                </span></span>
<span><span class="co">## [45] yaml_2.3.10                 codetools_0.2-20           </span></span>
<span><span class="co">## [47] curl_6.2.3                  lattice_0.22-7             </span></span>
<span><span class="co">## [49] tibble_3.2.1                plyr_1.8.9                 </span></span>
<span><span class="co">## [51] withr_3.0.2                 evaluate_1.0.3             </span></span>
<span><span class="co">## [53] desc_1.4.3                  BiocFileCache_2.99.5       </span></span>
<span><span class="co">## [55] xml2_1.3.8                  pillar_1.10.2              </span></span>
<span><span class="co">## [57] BiocManager_1.30.25         filelock_1.0.3             </span></span>
<span><span class="co">## [59] ncdf4_1.24                  RCurl_1.98-1.17            </span></span>
<span><span class="co">## [61] hms_1.1.3                   glue_1.8.0                 </span></span>
<span><span class="co">## [63] lazyeval_0.2.2              tools_4.5.0                </span></span>
<span><span class="co">## [65] data.table_1.17.4           QFeatures_1.19.2           </span></span>
<span><span class="co">## [67] mzR_2.43.0                  fs_1.6.6                   </span></span>
<span><span class="co">## [69] fastmatch_1.1-6             grid_4.5.0                 </span></span>
<span><span class="co">## [71] tidyr_1.3.1                 MsCoreUtils_1.21.0         </span></span>
<span><span class="co">## [73] cli_3.6.5                   rappdirs_0.3.3             </span></span>
<span><span class="co">## [75] textshaping_1.0.1           S4Arrays_1.9.1             </span></span>
<span><span class="co">## [77] dplyr_1.1.4                 AnnotationFilter_1.33.0    </span></span>
<span><span class="co">## [79] sass_0.4.10                 digest_0.6.37              </span></span>
<span><span class="co">## [81] SparseArray_1.9.0           htmlwidgets_1.6.4          </span></span>
<span><span class="co">## [83] memoise_2.0.1               htmltools_0.5.8.1          </span></span>
<span><span class="co">## [85] pkgdown_2.1.3.9000          lifecycle_1.0.4            </span></span>
<span><span class="co">## [87] httr_1.4.7                  bit64_4.6.0-1              </span></span>
<span><span class="co">## [89] MASS_7.3-65</span></span></code></pre>
</div>
  </main><aside class="col-md-3"><nav id="toc" aria-label="Table of contents"><h2>On this page</h2>
    </nav></aside>
</div>



    <footer><div class="pkgdown-footer-left">
  <p>Developed by Laurent Gatto, Johannes Rainer, Sebastian Gibb.</p>
</div>

<div class="pkgdown-footer-right">
  <p>Site built with <a href="https://pkgdown.r-lib.org/" class="external-link">pkgdown</a> 2.1.3.9000.</p>
</div>

    </footer>
</div>





  </body>
</html>
